internal/app/starter/master_linux_test.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package starter
+
+import (
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines"
+	"github.com/sylabs/singularity/internal/pkg/test"
+)
+
+func TestDoRPCThread(t *testing.T) {
+	test.DropPrivilege(t)
+	defer test.ResetPrivilege(t)
+
+	var fatal error
+	fatalChan := make(chan error, 1)
+
+	var fakeEngine engines.Engine
+
+	tests := []struct {
+		name         string
+		rpcSocket    int
+		containerPid int
+		engine       *engines.Engine
+		shallPass    bool
+	}{
+		{
+			name:         "invalid case",
+			rpcSocket:    -1,
+			containerPid: -1,
+			engine:       nil,
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine struct",
+			rpcSocket:    -1,
+			containerPid: -1,
+			engine:       &fakeEngine,
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine; fake rpcSocket",
+			rpcSocket:    42000,
+			containerPid: -1,
+			engine:       &fakeEngine,
+			shallPass:    false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			go doRPCthread(tt.rpcSocket, tt.containerPid, tt.engine, fatalChan)
+			fatal = <-fatalChan
```

Comment:
interesting... since the code does not have an obvious way to make it stop (IIRC there's a magic value that you can write to the RPC socket that will cause it to exit), it's not evident whether this test will hang here forever or not.

Maybe add a comment?

Reply:
Maybe add a comment?

Reply:
Let me try yet again: if I look at this test, it's not obvious that it will stop (i.e. that something is sent over `fatalChan`, partly because the name "fatalChan" suggest that would be the case only if a failure occurs).

Can you please add a comment clarifying that the test cases are such that there's always going to be a failure (both cases have `shallPass` set to `false`; someone reading this code might think there are other cases where `shallPass` is set to `true`) or that `createContainer` always sends something on `fatalChan` no matter what or whatever it is that guarantees that this test will not get stuck waiting for a value on `fatalChan`?